skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,213 @@
-        for coef in it_inner:
-            if coef == 0:
-                continue
-            tmp_pt_idx = np.add(b_lo, it_inner.multi_index)
-            tmp_pt_i = np.ravel_multi_index(tmp_pt_idx, mask.shape)
-
-            if mask[tuple(tmp_pt_idx)]:
-                matrix_unknown[mask_pt_n, tmp_pt_i] = coef
+        nvals = 0
+        for coef, i in zip(coefs, index1d):
+            if mask_flat[i]:
+                row_idx_unknown[idx_unknown] = mask_pt_n
+                col_idx_unknown[idx_unknown] = i
+                data_unknown[idx_unknown] = coef
+                idx_unknown += 1
             else:
-                matrix_known[mask_pt_n, tmp_pt_i] = coef
-
-    # Prepare diagonal matrix
-    flat_diag_image = sparse.dia_matrix((out.flatten(), np.array([0])),
-                                        shape=(out.size, out.size))
-
-    # Calculate right hand side as a sum of known matrix's columns
-    matrix_known = matrix_known.tocsr()
-    rhs = -(matrix_known * flat_diag_image).sum(axis=1)
+                data_known[idx_known, :] -= coef * out_flat[i, :]
+                nvals += 1
+        if nvals:
+            row_idx_known[idx_known] = mask_pt_n
+            idx_known += 1
+
+    # Call an efficient Cython-based implementation for all interior points
+    row_start = mask_pt_n + 1
+    known_start_idx = idx_known
+    unknown_start_idx = idx_unknown
+
+    nnz_rhs = _build_matrix_inner(
+        # starting indices
+        row_start, known_start_idx, unknown_start_idx,
+        # input arrays
+        center_i, raveled_offsets, coef_vals, mask_flat.astype(np.int16),
+        out_flat,
+        # output arrays
+        row_idx_known, data_known, row_idx_unknown, col_idx_unknown,
+        data_unknown
+    )
+
+    # trim RHS vector values and indices to the exact length
+    row_idx_known = row_idx_known[:nnz_rhs]
+    data_known = data_known[:nnz_rhs, :]
+
+    # Form sparse matrix of unknown values
+    sp_shape = (n_mask, out.size)
+    matrix_unknown = sparse.coo_matrix(
+        (data_unknown, (row_idx_unknown, col_idx_unknown)), shape=sp_shape
+    ).tocsr()
 
     # Solve linear system for masked points
     matrix_unknown = matrix_unknown[:, mask_i]
-    matrix_unknown = sparse.csr_matrix(matrix_unknown)
-    result = spsolve(matrix_unknown, rhs)
 
-    # Handle enormous values
-    result = np.clip(result, *limits)
+    # dense vectors representing the right hand side for each channel
+    rhs = np.zeros((n_mask, n_channels), dtype=out.dtype)
+    rhs[row_idx_known, :] = data_known
+
+    # set use_umfpack to False so float32 data is supported
+    result = spsolve(matrix_unknown, rhs, use_umfpack=False,
+                     permc_spec='MMD_ATA')
+    if result.ndim == 1:
+        result = result[:, np.newaxis]
 
-    result = result.ravel()
+    # Handle enormous values on a per-channel basis
+    for ch in range(n_channels):
+        result[..., ch] = np.clip(result[..., ch], *limits[ch])
```

Comment:
Thanks! We can also move this clipping to the outer `inpaint_biharmonic` function so it will only be called once instead of on each region individually.